check.go

```diff
@@ -0,0 +1,83 @@
+package main
+
+import (
+	"fmt"
+
+	gx "github.com/whyrusleeping/gx/gxutil"
+)
+
+type pkgImport struct {
+	parents []string
+	version string
+}
+
+func check(pkg *gx.Package) (bool, error) {
+	var failed bool
+	// name -> hash -> path+version
+	packages := make(map[string]map[string]*pkgImport)
+
+	var traverse func(*gx.Package) error
+	traverse = func(pkg *gx.Package) error {
+		return pkg.ForEachDep(func(dep *gx.Dependency, dpkg *gx.Package) error {
+			pkgset, ok := packages[dpkg.Name]
+			if !ok {
+				pkgset = make(map[string]*pkgImport, 1)
+				packages[dpkg.Name] = pkgset
+			}
+			imp, ok := pkgset[dep.Hash]
+			if !ok {
+				pkgset[dep.Hash] = &pkgImport{
+					version: dpkg.Version,
+					parents: []string{pkg.Name},
+				}
+				return traverse(dpkg)
+			}
+			imp.parents = append(imp.parents, pkg.Name)
+			return nil
+		})
+	}
+	if err := traverse(pkg); err != nil {
+		return !failed, err
+	}
+	for name, pkgset := range packages {
+		switch len(pkgset) {
+		case 0:
+			panic("can't have an empty package set")
```

Comment:
we will never have an empty package set? what if this is run on a package with no deps?

Reply:
Ah. Yeah, that was a truly horrible variable name. I've renamed it to `pkgVersions`. Basically, it's impossible to have fewer than one version (really, the check is kind of pointless).